### ISSUE TITLE
Boost-generated CLAUDE.md guidelines contain redundant/wasteful tokens

### ISSUE THREAD

**Issue by leek** [laravel/boost#606]
**Update:** I've decided to solve my own problem by using subdirectory-CLAUDE.md files, see comment: https://github.com/laravel/boost/issues/606#issuecomment-4451916468

---

> [!NOTE]
> This analysis was performed using Claude (Opus 4.6), so the "already in training data" observations are specific to Claude's knowledge. However, the duplication and verbosity issues apply universally, and most modern LLMs share similar baseline knowledge of Laravel/PHP conventions — so the savings would likely benefit other models too.

## Summary

The `<laravel-boost-guidelines>` injected into CLAUDE.md contain a significant amount of content that is either already in Claude's training data, duplicated across sections, or overly verbose. This burns context window tokens every conversation without improving output quality.

Rough estimate: **~220 of ~510 lines (~43%)** (in my personal generated file) could be removed or compressed without losing any behavioral benefit.

## 1. Content Claude Already Knows (Training Data)

These are standard Laravel/PHP conventions that don't need to be stated:

**PHP rules section:**
- "Always use curly braces for control structures" — standard, and already enforced by the Pint rule
- "Use PHP 8 constructor property promotion" — well-known since PHP 8.0
- "Always use explicit return type declarations" / "Use appropriate PHP type hints" — standard practice
- The code example showing a typed method signature — unnecessary illustration of basic PHP
- "Prefer PHPDoc blocks over inline comments" — generic best practice

**Laravel/core rules section:**
- "Use Eloquent models and relationships before raw queries"
- "Avoid `DB::`, prefer `Model::query()`"
- "Prevent N+1 with eager loading"
- "Use Form Request classes for validation"
- "Use gates, policies, Sanctum"
- "Prefer named routes and `route()`"
- "Use `ShouldQueue` for long operations"
- "Never use `env()` outside config files"
- "Use factories in tests"

**Livewire/core rules section** — entirely redundant. "Livewire allows you to build dynamic, reactive interfaces using only PHP" is explaining what Livewire is to a model that already knows. The only useful line ("activate the skill") is already stated in the Skills Activation section.

**Laravel 12 Structure** — the middleware/Kernel/bootstrap changes have been standard since Laravel 11 and are well within training data.

## 2. Content Duplicated Across Sections

| Instruction | Occurrences |
|---|---|
| "Use `search-docs` tool" | **5x** — boost rules, Laravel 12, Pest, Tailwind, Filament sections |
| "Activate `[skill-name]` skill" | **2x each** — Skills Activation section + individual package section (Livewire, Pest, Tailwind, AI SDK, Prism) |
| "Use `list-artisan-commands`" | **3x** — boost, laravel/core, filament sections |
| "Pass `--no-interaction`" | **2x** — laravel/core, filament sections |
| "Use `get-absolute-url`" | **2x** — boost, herd sections |
| Frontend bundling / Vite error advice | **2x** — foundation section, laravel/core section |

## 3. Verbose / Compressible Content

- **Boost search-docs syntax** (7 lines) — could be 1 line: "Supports stemming, AND logic, quoted phrases, and multiple queries"
- **Skills Activation descriptions** (~7 lines per skill) — each skill has a multi-line description of when to activate, but these descriptions are already present in the skill system prompts. A simple list of skill names would suffice.
- **Filament code snippets** (~100 lines) — helpful for v5, but the testing examples are standard Livewire testing patterns. These could be moved to a separate reference file rather than injected into every conversation. The **Common Mistakes** section (incorrect namespaces, breaking changes) is the actual high-value content here.

## Suggestions

1. **Remove standard PHP/Laravel knowledge** — trust the model's training data for basics; only include project-specific conventions or non-obvious choices
2. **Deduplicate** — state each instruction once (e.g., "use search-docs" at the top, skill activations in one place only)
3. **Move Filament code snippets** to a reference file that can be loaded on demand, keep only the Common Mistakes / namespace corrections inline
4. **Compress verbose descriptions** — the Skills Activation section could be just skill names since the skill prompts already contain activation triggers
5. **Delete the Livewire/core section** entirely — adds nothing beyond what's in training data + the skill activation already mentioned elsewhere

## Environment

- Laravel Boost v2
- Claude Code (Opus 4.6)
- Filament v5, Laravel 12, Livewire 4, Pest 4

**Comment by pushpak1300** [comment]
I agree with this. I think there is a real opportunity now to remove some of the redundant rules. Models have significantly improved and already have much of this knowledge built in, especially with the additional data we have added into the skill.

One important point is that this was measured on Cloud Opus 4.6, which is currently one of the strongest state of the art models. Our broader goal with Boost is to make it genuinely useful for smaller models as well, such as Kimi, GLM5, and others. We do not want it to be something that only benefits top tier models.

With something like Cloud Opus 4.6, you can often get comparable results even without Boost, simply because the model is already very capable. That said, I agree there is room to simplify and remove redundancy where it no longer adds value.

Thanks a lot for opening this issue. I will review the rules and clean things up accordingly.

**Comment by leek** [comment]
> One important point is that this was measured on Cloud Opus 4.6, which is currently one of the strongest state of the art models. Our broader goal with Boost is to make it genuinely useful for smaller models as well, such as Kimi, GLM5, and others. We do not want it to be something that only benefits top tier models.

I appreciate you taking a look and considering! 

Without adding too much complexity I wonder if it is possible to offer a "slim" install option or perhaps a specific output just for `CLAUDE.md`. Just thinking out loud...

Thanks for your work on a great tool!

**Comment by oskobri** [comment]
I completely agree with the points raised here. The CLAUDE.md file generated by Laravel Boost ends up far too verbose. In my view, most of what Boost currently injects into CLAUDE.md should not be there at all. These details should be moved into dedicated skills, each focused on a specific area (testing, Livewire, PHP, ...). CLAUDE.md should stay lightweight and contain only high‑level.

**Comment by iddo7** [comment]
I totally agree as well. On a brand new project with barely 2 small custom guidelines I get a warning from Claude Code saying the `CLAUDE.md` is too large and could hinder performance.

**Comment by carlosvaldesweb** [comment]
Agree

**Comment by MechJosh0** [comment]
Extending off [glennjacobs](https://github.com/glennjacobs) in https://github.com/laravel/boost/issues/738, he's suggested utilising `.claude/rules` but I also think making use of child directory `CLAUDE.md` as well.

**Comment by glennjacobs** [comment]
> Extending off [glennjacobs](https://github.com/glennjacobs) in [#738](https://github.com/laravel/boost/issues/738), he's suggested utilising `.claude/rules` but I also think making use of child directory `CLAUDE.md` as well.

Yeah you're right, it is the same problem. Rules are just a specific solution to it, instead of one massive CLAUDE.md where everything competes for attention, you split instructions into focused files that can be scoped to specific file types or directories, so they only load when relevant. Less noise, better context.

**Comment by glennjacobs** [comment]
It's worth noting Claude, Cursor and Codex all support rules now.

How do you envisage child directory `CLAUDE.md` files working?

I am coming from the perspective of a package developer primarily and wanted to offer rules, as the CLAUDE.md was massive when all merged in, and because of that a lot of it was getting missed/ignored.

**Comment by ItsReddi** [comment]
The idea of https://github.com/laravel/boost/issues/733, The idea stems from `best practice skill` and would give project owners the ability to determine how context is grouped.

**Comment by mishahawthorn** [comment]
Would also propose processing any included guideline content through a token "minifier" pipeline like https://github.com/JuliusBrussee/caveman. Fewer tokens used and purportedly better task completion accuracy overall.

**Comment by youaresodalgona** [comment]
Agree, Is rule-based support not planned?

**Comment by lukebouch** [comment]
I would agree with everyone’s comments here. I think the CLAUDE.md is too verbose and causes the context to fill up way to fast resulting in a net decrease in performance.

**Comment by kngbwsr** [comment]
After hours of programming skills/rules, here's my take and modified suggestions:

**1. Keep standard PHP/Laravel knowledge** — 
You cannot trust a model, you cannot update a model. You can trust established rules set by the devs. This lets additional models that may not be as smart work better. 

**2. Deduplicate** — 
yes, but, while dedup'ing also replace commonly used commands, instructions, etc with a core "Laravel Skill" that has it's own skill file with nested documentation references that the skill can draw on during implementation depending on the task itself.  I programmed one of these using the Laravel docs, it works amazingly during planning.

**3. Move Filament code** — 
For the love of all that is holy yes... why in the world did filament think they needed to inject their details into the guidelines? It works perfect as a skill.... I have actually created a separate skill that does just this. the main skill md file, with the reference documentation. The skill file being a master orchestrator and info gatherer. While this works now, the next filament update will require a revamp of the skill... and that's a hassle.

**4. Compress verbose descriptions** — 
100%, because of compression most llms only read the first lines, the last lines, and summarize the middle of most the injected context.  If you optimize for summary the context will keep 100% of that already summarized data
**5. Delete the Livewire/core section entirely** — create a separate livewire skill.

I have literally counted 400 lines of code that are just injected package guidelines that have no business being there. yes, this is the package creators choice, but boost could simply fix it for the good of all as well.

Extra documentation is not the issue, the issue is making the agent md the big brain, when it's actually the guardrail and orchestrator. 

I've literally replaced nearly all the guidelines with skills already, but if Laravel would simply not allow guidelines to be injected, and reverted them to Skills/Rules, and then generate a referencing skill-tree instead, we'd all be much better off.

**Comment by kngbwsr** [comment]
I actually just published a package today, still in dev, that handles a lot of this.  feel free to review, use, pr etc. It's my first package publication (although I've been coding for a long time). I don't usually share stuff, but this is something I think the community could use:
https://github.com/kngbwsr/laravel-agent-optimizer
Full disclosure, I just uploaded it, and haven't installed it myself yet as I'm super busy, but it's based on code that i've used in my project.  I just made a package from it. So, there will most likely be bugs.

Cheers

**Comment by leek** [comment]
I prefer **subdirectory scoped** CLAUDE.md files (ie: `app\Models\CLAUDE.md`, etc) so I built this:

Scoped CLAUDE.md/AGENTS.md next to every Laravel dir that matters — Models, Controllers, Form Requests, Resources, Policies, Jobs, Livewire, Notifications, Pennant, migrations, Pest tests, more.

_Use it **with** Laravel Boost if you want..._

```bash
npx apply-agent-rules apply leek/laravel-agent-rules
```

Ref: [github.com/leek/laravel-agent-rules](https://github.com/leek/laravel-agent-rules)

**Comment by kngbwsr** [comment]
@leek I just starred all your projects, they are 100% going into my agent flow.  I have actually been developing something similar leveraging the filesystem itself, but it looks like you already have it figured out.  There's one more thing I don't need to do haha.

This approach is far superior generally, but doesn't fix the boost bloat to the root agent file from what I can see? The package I set up reduces that considerably allowing you to use vendor provided guidelines without the bloat.  Using that, and your packages combined would meaningfully reduce context. 

I'm excited to implement your packages into my agent flow.

**Comment by leek** [comment]
> This approach is far superior generally, but doesn't fix the boost bloat to the root agent file from what I can see?

Yeah, I deliberately excluded any root level agent files so that I could continue to work with Boost. I will definitely give your optimizer a try as well. Thanks!

**Comment by dac514** [comment]
> Extending off [glennjacobs](https://github.com/glennjacobs) in [#738](https://github.com/laravel/boost/issues/738), he's suggested utilising `.claude/rules` but I also think making use of child directory `CLAUDE.md` as well.

Firstly, I agree with the general idea that putting everything in root `CLAUDE.md` file is not ideal. 

There have been more interesting proposals, complicated even, than the one I am about to propose myself, but I'd simply like to be able to control the path where `<laravel-boost-guidelines />` get written. 

For me, a config in `boost.json` that lets me  redirect all the output to `.claude/rules/laravel-boost.md` is a good enough solve for now.

**Comment by pushpak1300** [comment]
yes we are trying to solve it check #876

**Comment by CicerBro** [comment]
As a side note: According to Cursor, it reads all these files `.cursor/rules/laravel-boost.mdc`, `AGENTS.md` and `CLAUDE.md` which blows up the context window massively. I have all 3 enabled in Boost due to switching Codex, Cursor & Claude often.

That is ~45 KB of overlapping instructions on every request. Is this fixable somehow? Perhaps Boost should add these to `.cursorignore`

**Comment by pushpak1300** [comment]
> As a side note: According to Cursor, it reads all these files .cursor/rules/laravel-boost.mdc, AGENTS.md and CLAUDE.md which blows up the context window massively. I have all 3 enabled in Boost due to switching Codex, Cursor & Claude often.
> 
> That is ~45 KB of overlapping instructions on every request. Is this fixable somehow? Perhaps Boost should add these to .cursorignore
> 

what ? 😕

**Comment by pushpak1300** [comment]
so if my project has both AGENTS.md and CLAUDE.md it will use both files ?

Seems issue on harness side.

**Comment by iddo7** [comment]
> so if my project has both AGENTS.md and CLAUDE.md it will use both files ?
> 
> Seems issue on harness side.

I am guessing the model itself chooses to read both, not the harness

**Comment by bdlowery** [comment]
> so if my project has both AGENTS.md and CLAUDE.md it will use both files ?
> 
> Seems issue on harness side.

If people are freshly installing boost without an AGENTS.md and CLAUDE.md, boost could default to using AGENTS.md and do this for CLAUDE.md (the `@AGENTS.md` link) - https://code.claude.com/docs/en/memory#agents-md.

If they already have a combination of AGENTS.md / CLAUDE.md files existing there could be a prompt in the installer about it.

iirc, Anthropic is the only ai provider who refuses to use the AGENTS.md standard.

**Comment by CicerBro** [comment]
> > As a side note: According to Cursor, it reads all these files .cursor/rules/laravel-boost.mdc, AGENTS.md and CLAUDE.md which blows up the context window massively. I have all 3 enabled in Boost due to switching Codex, Cursor & Claude often.
> > That is ~45 KB of overlapping instructions on every request. Is this fixable somehow? Perhaps Boost should add these to .cursorignore
> 
> what ? 😕

What was unclear about my comment?
